Add BMP280 device binding

chaudi opened the pull request:
Only I2C implementation for now to see if I'm on the right track.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=105) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=105) before we can accept your contribution.<br/><br/>:x: chaudi [sign now](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=105)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/iot?pullRequest=105) it.</sub>

joperezr commented on src/devices/README.md:
```diff
@@ -7,6 +7,7 @@ Our vision: the majority of .NET bindings are written completely in .NET languag
 ## Binding Index
 
 * [Mcp3008](Mcp3008/README.md)
+* [Bmp280](Bmp280/README.md)
```
NIT: Can we order these alphabetically?

joperezr commented on src/devices/Bmp280/Bmp280.sln:
```diff
@@ -0,0 +1,36 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
```
We are not checking in solutions for every binding just to avoid having too many solutions in the project that will slow down nuget restore.

joperezr commented on src/devices/Bmp280/IBmp280.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Iot.Device
+{
+    public interface IBmp280
```
Why do you need an interface here? I don't think it is very useful to implement it since it would be the only class implementing it anyway so I would just remove this file.

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,592 @@
+        private enum CommunicationProtocol
+        {
+            Gpio,
+            Spi,
```
NIT: I would only add the I2c Communication Protocol for now on this enum, and add the Spi and Gpio ones once we add support for them.

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,592 @@
+    public class Bmp280 : IDisposable, IBmp280
+    {
+        private I2cDevice _i2cDevice;
+        private CommunicationProtocol protocol;
```
`_communicationProtocol` instead of `protocol`

chaudi replied:
fixed

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,592 @@
+    {
+        private I2cDevice _i2cDevice;
+        private CommunicationProtocol protocol;
+        private bool initialised = false;
```
`_initialized` instead of `initialised`

chaudi replied:
fixed

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,592 @@
+        private I2cDevice _i2cDevice;
+        private CommunicationProtocol protocol;
+        private bool initialised = false;
+        private Bmp280CalibrationData CalibrationData;
```
`_calibrationData` Same feedback for all of your private fields.

chaudi replied:
fixed

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,592 @@
+        private CommunicationProtocol protocol;
+        private bool initialised = false;
+        private Bmp280CalibrationData CalibrationData;
+        protected readonly byte Signature = 0x58;
```
Should this be private? why do we want it to be protected?

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,592 @@
+        private bool initialised = false;
+        private Bmp280CalibrationData CalibrationData;
+        protected readonly byte Signature = 0x58;
+        private int TFine;
```
I think we should rename this as it is not clear what TFine means

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,592 @@
+            protocol = CommunicationProtocol.I2c;
+        }
+
+        public Bmp280(I2cConnectionSettings settings)
```
I would remove this constructor. I get that what you are trying to do is to make this work on Linux and Windows without changing the consuming code, but on past device bidnings we have opted to just provide the constructor that takes in the I2cDevice

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,592 @@
+        /// <summary>
+        /// Sets mode to forced (device goes to sleep after one reading)
+        /// </summary>
+        public void SetModeForced()
```
Is this how the mode is called in the documentation? I just ask because it is not super clear to me what forced means until I read your comment.

chaudi replied:
Thanks for all feedback!

It is called forced mode everywhere in the datasheet:
> BMP280 can be operated in three power modes (see chapter 3.6):
> sleep mode
> normal mode
> forced mode

joperezr replied:
I see, in that case this is fine. Having the summary comments helps a bunch, since consumers that don't know what it means (like me :smile:) will be able to know what is the use for it.

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,562 @@
+        private void Begin()
+        {
+            _i2cDevice.WriteByte((byte)Registers.REGISTER_CHIPID);
+            var readSignature = _i2cDevice.ReadByte();
```
In general, we try to avoid the use of `var` as much as possible since it might make the code less readable. One of the very few exceptions is when you are immediatly assigning it using the new keyword so the type is super obvious, like `var myType = new MyType();`. This feedback applies to the rest of your usage of `var`.

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,562 @@
+            _i2cDevice.Write(new[] { (byte)Registers.REGISTER_CONTROL, status });
+        }
+
+        public PowerMode ReadPowerMode()
```
>PowerMode [](start = 15, length = 9)

Should this just be named `ReadMode()` that way it will echo our `SetMode` methods. If not, then we could change the `SetMode` ones for `SetPowerMode` instead.

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,562 @@
+        }
+
+        private Sampling ByteToSampling(byte value)
+        {
```
Given that you are controlling all of the callsites for this method, should you simplify this to something like:
```c#
private Sampling ByteToSampling(byte value) => (Sampling)value;
```

chaudi replied:
According to the datasheet values >5 are ultra high resolution so just casting to Sampling won't return the correct Sampling value for all cases.

Is something like this preferred?

```C#
private Sampling ByteToSampling(byte value)
{
    //Values >=5 equals UltraHighResolution
    if (value >= 5)
    {
        return Sampling.UltraHighResolution;
     }
     return (Sampling) value;            
}
```

joperezr replied:
sure, that sounds good to me since, as I said earlier, you control all the call sites so you know that you won't get some weird input passed in.

chaudi replied:
Just want to be sure to return the correct sampling value is returned if a higher value is read from the device.

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,562 @@
+        /// <returns>
+        ///  Degrees celsius
+        /// </returns>
+        private double BMP280_compensate_T_double(int adcT)
```
Even though these two compensate methods are not part of the public surface area, I would try to name them a bit more readable, using PascalCasing instead of _ and using full words instead of just T for temperature or p for preassure

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,562 @@
+        /// <returns>
+        ///  Value from register
+        /// </returns>
+        public byte Read8(byte register)
```
I would make these methods private, as they are only really used inside your binding, and might just confuse consumers when they are looking for the higher level methods like ReadTemperature or ReadPreassure. Also, I know that this is how they were named in Adafruit's repo, but can we rename them to either `ReadByte` `ReadShort` or at least something like `Read8BitsFromRegister` `Read16BitsFromRegister`? That will just make it feel more like .NET and more close according to the framework.

chaudi replied:
Fixed

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,562 @@
+        /// <summary>
+        /// BMP280s power modes
+        /// </summary>
+        public enum PowerMode : byte
```
I would add one extra name in your namespace, (Iot.Device.Bmp280 instead of just Iot.Device) and then move these three enums to their own individual files in order to get these class a bit more clean. That is also the pattern that we have been following with these device bindings.

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,455 @@
+        private void ReadCoefficients()
+        {
+            // Read temperature calibration data
+            _calibrationData.DigT1 = Read16FromRegister((byte)Registers.REGISTER_DIG_T1);
```
By reading the datasheet, shouldn't some of these be unsigned and some of them signed? I also think that some of them should be interpreted as little endian and some as big endian but I'm not 100% sure.

chaudi replied:
I got data that seemed reasonable when I tested but I'll double check.

joperezr replied:
I only mentioned this because we used to have a very similar binding for a bme280 here: https://github.com/dotnet/corefxlab/blob/master/samples/System.Devices.Gpio.Samples/PressureTemperatureHumiditySensor.cs  and in there I remember that we needed to take into account that some were signed, and some were not.

chaudi replied:
Only the P1 and T1 are unsigned so it seems to be correct. Changed from int/uint to short instead.

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,455 @@
+                _i2cDevice.Read(readBuffer);
+                int h = readBuffer[1] << 8;
+                int l = readBuffer[0];
+                ushort value = (ushort)(h + l);
```
Looks like you are doing the Endianess yourself here which might be tricky and not work all of the time. Have you tried using [BinaryPrimitives](https://github.com/dotnet/corefx/blob/master/src/System.Memory/ref/System.Memory.cs#L256) class?

chaudi replied:
Now I have, thanks.

chaudi commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,592 @@
+        /// <summary>
+        ///  Reads the factory-set coefficients 
+        /// </summary>
+        private void ReadCoefficients()
```
Hm, I'm equivalent about where to place it so I moved it in the latest check in.

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,420 @@
+        {
+            if (_communicationProtocol == CommunicationProtocol.I2c)
+            {
+                byte[] readBuffer = { 0x00, 0x00 };
```
NIT: if you change the `LangVersion` property in your .csproj, you'll be able to change this to instead be: `Span<byte> bytes = stackalloc byte[2];` which will be more performant. You also don't need to initialize the buffer to 0s.

chaudi replied:
Fixed.

joperezr commented on src/devices/Bmp280/Bmp280.cs:
```diff
@@ -0,0 +1,420 @@
+                _i2cDevice.WriteByte(register);
+                _i2cDevice.Read(readBuffer);
+
+                return BinaryPrimitives.ReadUInt32LittleEndian(readBuffer);
```
I'm not 100% sure but I believe this will throw because readBuffer doesn't have 4 bytes which is what an Int32 requires. In order to satisfy this, you could have your readBuffer be of size 4, and just have the readBuffer[0] to be 0x00 and start the read from readBuffer[1].

chaudi replied:
Fixed

joperezr commented:
Left two more small comments, but this is almost ready to go in! Thanks a lot for addressing all of the comments!

chaudi commented:
> 
> 
> Left two more small comments, but this is almost ready to go in! Thanks a lot for addressing all of the comments!

Thanks a lot for all the feedback!
I'll keep them in mind for eventual future PRs.
